### ISSUE TITLE
[error] Error / :vehicle_not_found

### ISSUE THREAD

**Issue by ngardiner** [adriankumpf/teslamate#181]
**Describe the bug**

After running overnight, I found the error message below in the teslamate logs, and saw that the teslamate web interface was stuck attempting to refresh.

Given I won't be monitoring the output of teslamate daily, could this error trigger a reconnect/restart/retry internally rather than requiring a manual restart?

**To Reproduce**
1. Start teslamate
2. Let it run for 24 hours
3. Review logs, identify if this error is present

**Relevant entries from the logs**

This morning, I checked the status of teslamate to find that throughout the night, it seems to have stopped tracking my Tesla with this error:

```
23:30:35.499 car_id=1 [info] Suspending logging
23:51:46.943 car_id=1 [info] Start / :offline
01:52:24.996 car_id=1 [error] Error / :vehicle_not_found
01:53:25.177 car_id=1 [error] Error / :vehicle_not_found
01:54:25.392 car_id=1 [error] Error / :vehicle_not_found
[...]
08:24:07.427 car_id=1 [error] Error / :vehicle_not_found
08:25:07.621 car_id=1 [error] Error / :vehicle_not_found
```

The issue is, I had to manually restart teslamate for the issue to resolve:

```
08:26:25.980 [info] Running TeslaMateWeb.Endpoint with cowboy 2.6.3 at ::
:4000 (http)
08:26:26.011 [info] Access TeslaMateWeb.Endpoint at http://teslamate.[domain]
08:26:27.007 [info] Found '449107'
08:26:27.026 [info] Initializing handler
08:26:27.026 [info] Connection has been established
08:26:27.260 car_id=1 [info] Start / :offline
08:26:55.494 [info] GET /
08:26:55.511 [info] Sent 200 in 16ms
08:26:57.928 [info] GET /
08:26:57.934 [info] Sent 200 in 5ms
08:27:01.213 [info] GET /
08:27:01.220 [info] Sent 200 in 6ms
```

**Operating environment (please complete the following information):**
 - OS: Ubuntu Disco (19.04)
 - Version v1.9.1-3-gf43e016

**Comment by SigLinJo** [comment]
Had a similar problem. Came home after a really busy day with the car to notice that Grafana had lost over 24 hour of data, check Teslamate web interface and it is stuck in charging from yesterday 00:30 something and my logs are filled with this, a row every time teslamate tried to query my car.

02:52:54.386 car_id=1 [warn] Error / :unknown

A manual restart of TeslaMate restored functionality, would love for TeslaMate to internally recognize something like that and restart, or if not possible maybe set up some way to send a warning that something is wrong.

I lost some valuable data today, all on me of course for not checking in on teslamate during the day, but anything that can stop it from happening in the future would be wonderful!

I'm running a manual install in a FreeBSD jail on FreeNAS, and TeslaMate 1.9.1 and have a Tesla Model 3 SR+ in Sweden with firmware 2019.32.2.1 9b8d6cd installed

Still grateful for all the functionality of the program, just wanted to add in here that I also noticed the problem, although I had an other error or warning in my logs.

**Comment by ngardiner** [comment]
Just to help with correlation, I am also an SR+ and am on 2019.32.2.2 and my installation is manual on Ubuntu.

**Comment by adriankumpf** [comment]
I had the same problem yesterday and had to restart. Tesla have changed the IDs of some cars again. This isn't the first time this has happened: #142, #136, #90. But this time the change was handled properly by TM without crashing. However, a manual restart is still required. I hope to change that with the next release.

**Comment by ngardiner** [comment]
Got it. Is it possible to just re-query the Tesla API for the new vehicle ID when the one you were using previously reports that it is invalid? I assume that's what's happened when the process was restarted?

Given the low likelihood that a vehicle that was operating previously is now invalid on the next polling cycle, it seems safe enough logic?

**Comment by adriankumpf** [comment]
In theory, yes. It's a little more complicated than that though. Mainly because I didn't built TM with the assumption that IDs may change. Furthermore, the Tesla servers sometimes return no vehicles at all due to backend errors. Those kinds of errors need to be handled differently that ID changes.

**Comment by bahree** [comment]
I wonder if we can run a sql script that can update the ID's across the board - fetch the new one, and disable constraints - update it, and then go back and enable it? Might be something to run via the settings screen and not all the time. Thoughts?

Also I poked around in the data returned with the new v10 update (and I don't know when this started to be clear) - but from my interpretation of this, the ID's can change more. See below - I have retracted some details but the `id_s` seems like a collection which could have more than one association.

`"response": {
        "id": 3333333333333333,
        "vehicle_id": 1234567890,
        "vin": "MYVINTHATIAMNOTSHARING",
        "display_name": "Squirty",
        "option_codes": "AD15,MDL3,PBSB,RENA,BT37,ID3W,RF3G,S3PB,DRLH,DV2W,W39B,APF0,COUS,BC3B,CH07,PC30,FC3P,FG31,GLFR,HL31,HM31,IL31,LTPB,MR31,FM3B,RS3H,SA3P,STCP,SC04,SU3C,T3CA,TW00,TM00,UT3P,WR00,AU3P,APH3,AF00,ZCST,MI00,CDM0",
        "color": null,
        "tokens": [
            "f2ff88888f943ff",
            "f2ff88888f943ff3434"
        ],
        "state": "online",
        "in_service": false,
        "id_s": "3333333333333333",
        "calendar_enabled": true,
        "api_version": 6,
        "backseat_token": null,
        "backseat_token_updated_at": null
    }`

Thanks.

**Comment by jun3280net** [comment]
Hi @adriankumpf  - getting similar issue with the V10 update.  Tried restarting Teslamate, but getting the following error : 

teslamate_1  | 09:39:53.056 [info] Already up
teslamate_1  | 09:40:05.499 [warn] The persisted API tokens could not be refreshed. Please sign in again.
teslamate_1  | %TeslaApi.Error{env: %Tesla.Env{__client__: %Tesla.Client{adapter: {Tesla.Adapter.Hackney, :call, [[recv_timeout: 30000]]}, fun: nil, post: [], pre: [{Tesla.Middleware.BaseUrl, :call, ["https://owner-api.teslamotors.com/"]}, {Tesla.Middleware.Headers, :call, [[{"user-agent", "github.com/mgwidmann/tesla_api"}]]}, {Tesla.Middleware.JSON, :call, [[]]}, {Tesla.Middleware.Headers, :call, [[{"Authorization", "Bearer **XXXNUM**"}]]}]}, __module__: Tesla, body: %{"error" => "invalid_grant", "error_description" => "The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}, headers: [{"server", "nginx"}, {"date", "Tue, 01 Oct 2019 15:40:05 GMT"}, {"content-type", "application/json; charset=utf-8"}, {"transfer-encoding", "chunked"}, {"connection", "keep-alive"}, {"x-frame-options", "SAMEORIGIN"}, {"x-xss-protection", "1; mode=block"}, {"x-content-type-options", "nosniff"}, {"cache-control", "no-store"}, {"pragma", "no-cache"}, {"www-authenticate", "Bearer realm=\"Doorkeeper\", error=\"invalid_grant\", error_description=\"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.\""}, {"x-request-id", "**NOTSUREIFIMPORTANT**"}, {"x-runtime", "0.007159"}], method: :post, opts: [], query: [], status: 401, url: "https://owner-api.teslamotors.com/oauth/token"}, error: :authentication_failure, message: "Failed to authenticate."}

teslamate_1  | 09:40:05.511 [info] Running TeslaMateWeb.Endpoint with cowboy 2.6.3 at :::4000 (http)
teslamate_1  | 09:40:05.516 [info] Access TeslaMateWeb.Endpoint at http://192.168.14.129
teslamate_1  | 09:40:05.529 [info] Found car ''
teslamate_1  | 09:40:05.555 car_id=1 [warn] Error / :not_signed_in
mosquitto_1  | 1569944405: New connection from 172.18.0.2 on port 1883.
mosquitto_1  | 1569944405: New client connected from 172.18.0.2 as TESLAMATE_**NOTSUREIFIMPORTANT** (p2, c1, k60).
teslamate_1  | 09:40:05.564 [info] Initializing handler
teslamate_1  | 09:40:05.567 [info] Connection has been established
teslamate_1  | 09:40:15.557 car_id=1 [warn] Error / :not_signed_in

**Comment by adriankumpf** [comment]
That's a different error, isn't it? Like it says "The persisted API tokens could not be refreshed. Please sign in again." I can hardly imagine it just happening like that. Your tokens were invalidated at some point. For example, because of a password change or if other services that have your tokens do a token refresh.

**Comment by jun3280net** [comment]
> That's a different error, isn't it? Like it says "The persisted API tokens could not be refreshed. Please sign in again." I can hardly imagine it just happening like that. Your tokens were invalidated at some point. For example, because of a password change or if other services that have the tokens refreshed them.

hi @adriankumpf.  How do I re-login?  I haven't changed any password, and aside from Teslamate - I don't think there are any other services that use the tokens. On a separate note, I did a docker-compose pull.....and now when I try and run Teslamate, I get the following :

:00.632 [info] Access TeslaMateWeb.Endpoint at http://192.168.14.50
teslamate_1  | 21:30:00.677 [info] Adding elevation to 512 positions ...
teslamate_1  | 21:30:00.728 [info] Querying location for elevation takes longer than 50ms ...
teslamate_1  | 21:30:01.085 [warn] Please sign in again.
teslamate_1  | 
teslamate_1  | %TeslaApi.Error{
teslamate_1  |   env: %Tesla.Env{
teslamate_1  |     __client__: %Tesla.Client{
teslamate_1  |       adapter: {Tesla.Adapter.Hackney, :call, [[recv_timeout: 30000]]},
teslamate_1  |       fun: nil,
teslamate_1  |       post: [],
teslamate_1  |       pre: [
teslamate_1  |         {Tesla.Middleware.BaseUrl, :call,
teslamate_1  |          ["https://owner-api.teslamotors.com/"]},
teslamate_1  |         {Tesla.Middleware.Headers, :call,
teslamate_1  |          [[{"user-agent", "github.com/mgwidmann/tesla_api"}]]},
teslamate_1  |         {Tesla.Middleware.JSON, :call, [[]]},
teslamate_1  |         {Tesla.Middleware.Headers, :call,
teslamate_1  |          [
teslamate_1  |            [
teslamate_1  |              {"Authorization",
teslamate_1  |               "Bearer xxx"}
teslamate_1  |            ]
teslamate_1  |          ]}
teslamate_1  |       ]
teslamate_1  |     },
teslamate_1  |     __module__: Tesla,
teslamate_1  |     body: %{
teslamate_1  |       "error" => "invalid_grant",
teslamate_1  |       "error_description" => "The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."
teslamate_1  |     },
teslamate_1  |     headers: [
teslamate_1  |       {"server", "nginx"},
teslamate_1  |       {"date", "Wed, 02 Oct 2019 03:30:01 GMT"},
teslamate_1  |       {"content-type", "application/json; charset=utf-8"},
teslamate_1  |       {"transfer-encoding", "chunked"},
teslamate_1  |       {"connection", "keep-alive"},
teslamate_1  |       {"x-frame-options", "SAMEORIGIN"},
teslamate_1  |       {"x-xss-protection", "1; mode=block"},
teslamate_1  |       {"x-content-type-options", "nosniff"},
teslamate_1  |       {"cache-control", "no-store"},
teslamate_1  |       {"pragma", "no-cache"},
teslamate_1  |       {"www-authenticate",
teslamate_1  |        "Bearer realm=\"Doorkeeper\", error=\"invalid_grant\", error_description=\"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.\""},
teslamate_1  |       {"x-request-id", "yyy"},
teslamate_1  |       {"x-runtime", "0.007053"}
teslamate_1  |     ],
teslamate_1  |     method: :post,
teslamate_1  |     opts: [],
teslamate_1  |     query: [],
teslamate_1  |     status: 401,
teslamate_1  |     url: "https://owner-api.teslamotors.com/oauth/token"
teslamate_1  |   },
teslamate_1  |   error: :authentication_failure,
teslamate_1  |   message: "Failed to authenticate."
teslamate_1  | }
teslamate_1  | 21:30:01.095 [info] Found ''
teslamate_1  | 21:30:01.114 [info] Application teslamate exited: TeslaMate.Application.start(:normal, []) returned an error: shutdown: failed to start child: TeslaMate.Vehicles
teslamate_1  |     ** (EXIT) an exception was raised:
teslamate_1  |         ** (ArgumentError) nil given for :vin. Comparison with nil is forbidden as it is unsafe. Instead write a query with is_nil/1, for example: is_nil(s.vin)
teslamate_1  |             (ecto) lib/ecto/query/builder/filter.ex:128: Ecto.Query.Builder.Filter.kw!/7
teslamate_1  |             (ecto) lib/ecto/query/builder/filter.ex:121: Ecto.Query.Builder.Filter.kw!/3
teslamate_1  |             (ecto) lib/ecto/query/builder/filter.ex:103: Ecto.Query.Builder.Filter.filter!/6
teslamate_1  |             (ecto) lib/ecto/query/builder/filter.ex:115: Ecto.Query.Builder.Filter.filter!/7
teslamate_1  |             (ecto) lib/ecto/repo/queryable.ex:40: Ecto.Repo.Queryable.get_by/4
teslamate_1  |             (teslamate) lib/teslamate/vehicles.ex:72: TeslaMate.Vehicles.create_or_update!/1
teslamate_1  |             (teslamate) lib/teslamate/vehicles.ex:26: anonymous fn/2 in TeslaMate.Vehicles.init/1
teslamate_1  |             (elixir) lib/enum.ex:1336: Enum."-map/2-lists^map/1-0-"/2
teslamate_1  | {"Kernel pid terminated",application_controller,"{application_start_failure,teslamate,{{shutdown,{failed_to_start_child,'Elixir.TeslaMate.Vehicles',{#{'__exception__' => true,'__struct__' => 'Elixir.ArgumentError',message => <<\"nil given for :vin. Comparison with nil is forbidden as it is unsafe. Instead write a query with is_nil/1, for example: is_nil(s.vin)\">>},[{'Elixir.Ecto.Query.Builder.Filter','kw!',7,[{file,\"lib/ecto/query/builder/filter.ex\"},{line,128}]},{'Elixir.Ecto.Query.Builder.Filter','kw!',3,[{file,\"lib/ecto/query/builder/filter.ex\"},{line,121}]},{'Elixir.Ecto.Query.Builder.Filter','filter!',6,[{file,\"lib/ecto/query/builder/filter.ex\"},{line,103}]},{'Elixir.Ecto.Query.Builder.Filter','filter!',7,[{file,\"lib/ecto/query/builder/filter.ex\"},{line,115}]},{'Elixir.Ecto.Repo.Queryable',get_by,4,[{file,\"lib/ecto/repo/queryable.ex\"},{line,40}]},{'Elixir.TeslaMate.Vehicles','create_or_update!',1,[{file,\"lib/teslamate/vehicles.ex\"},{line,72}]},{'Elixir.TeslaMate.Vehicles','-init/1-fun-0-',2,[{file,\"lib/teslamate/vehicles.ex\"},{line,26}]},{'Elixir.Enum','-map/2-lists^map/1-0-',2,[{file,\"lib/enum.ex\"},{line,1336}]}]}}},{'Elixir.TeslaMate.Application',start,[normal,[]]}}}"}
teslamate_1  | Kernel pid terminated (application_controller) ({application_start_failure,teslamate,{{shutdown,{failed_to_start_child,'Elixir.TeslaMate.Vehicles',{#{'__exception__' => true,'__struct__' => 'Elixir.Ar
teslamate_1  | 
teslamate_1  | Crash dump is being written to: erl_crash.dump...done
db_1         | 2019-10-02 03:30:04.762 UTC [53] LOG:  incomplete startup packet

**Comment by adriankumpf** [comment]
Did you try to sign in  http://192.168.14.50:4000?

**Comment by jun3280net** [comment]
> Did you try to sign in at http://192.168.14.50:4000?

Hi @adriankumpf  - I tried going to the page.    It says This site can't be reached.  It's probably related to the "crash_dump"

**Comment by adriankumpf** [comment]
Ah, okay got it. Could you open an new issue with the crash dump? This seems to be something different.

**Comment by jun3280net** [comment]
> Ah, okay, got it. Could you open an new issue with the crash dump? This seems to be something different.

@adriankumpf no prob.  I'll copy what I posted in my last message into a new issue.  Thanks!

**Comment by adriankumpf** [comment]
Fixed with 609e73e
